python/test_util.py

```diff
@@ -0,0 +1,221 @@
+                        str(Time.from_seconds(seconds, precision)), expected,
+                        msg.format(seconds))
+            check(2, fixed_2, '{} fixed 2')
+            check(1, fixed_1, '{} fixed 1')
+            check(0, fixed_0, '{} fixed 0')
+
+    def test_overlay_lists(self):
+        data = (
+            (0, ('---', '+--', '++-', '+++', '+++')),
+            (1, ('---', '-+-', '-++', '-++', '-++')),
+            (5, ('---', '---', '---', '---', '---')),
+            (-1, ('---', '---', '+--', '++-', '+++')),
+            (-4, ('---', '---', '---', '---', '---')),
+        )
+        for offset, expecteds in data:
+            for i, expected in enumerate(expecteds):
+                base = ['-'] * 3
+                new = ['+'] * i
+                overlay_lists(base, new, offset)
+                self.assertEqual(base, list(expected))
+
+    def test_caching_property(self):
+        self.computed = 0
+        class Foo:
+            def __init__(self, return_value):
+                self.return_value = return_value
+
+            @property
+            @cache
+            def something(instance):
+                self.computed += 1
+                return instance.return_value
+
+        f1 = Foo('bananas')
+        self.assertEqual(f1.something, 'bananas')
+        self.assertEqual(self.computed, 1)
+        self.assertEqual(f1.something, 'bananas')
+        self.assertEqual(self.computed, 1)
+        f2 = Foo('')
+        self.assertEqual(f2.something, '')
+        self.assertEqual(self.computed, 2)
+        self.assertEqual(f2.something, '')
+        self.assertEqual(self.computed, 2)
+
+
+class TestAB(TestCase):
+    def test_variable_constructor(self):
+        self.assertEqual(AB(1), AB(1, 1))
+        self.assertRaises(TypeError, AB, 1, 2, 3)
+
+    def test_eq(self):
+        self.assertEqual(AB(1, 2), AB(1, 2))
+        self.assertEqual(AB(1, 2), (1, 2))
+        self.assertNotEqual(AB(1, 2), AB(3, 4))
+        self.assertNotEqual(AB(1, 2), (1, 2, 3))
+        self.assertEqual(AB(1, 2) + AB(3, 2), AB(4, 4))
+        self.assertEqual(AB(1, 2) - AB(3, 2), AB(-2, 0))
+
+    def test_get_set_item(self):
+        chars = 'ab'
+        ab = AB(*chars)
+        for i, c in enumerate(chars):
+            self.assertEqual(ab[i], c)
+            self.assertEqual(getattr(ab, c), c)
+            ab[i] = i
+            self.assertEqual(ab[i], i)
+
+    def test_reversed(self):
+        self.assertEqual(AB(1, 2).reversed(), AB(2, 1))
+
+    def test_from_map(self):
+        plus_one = lambda n: n + 1
+        self.assertEqual(AB.from_map(plus_one, AB(1, 2)), AB(2, 3))
+        self.assertRaises(TypeError, AB.from_map, plus_one, (1, 2, 3))
+
+    def test_map(self):
+        times_two = lambda n: n * 2
+        ab = AB(1, 2)
+        self.assertEqual(ab.map(times_two), AB(2, 4))
+        times_m = lambda n, m: n * m
+        self.assertEqual(ab.map(times_m, 3), AB(3, 6))
+
+    def test_attr_lookup_and_call(self):
```

Comment:
`AB`s have no `index` method! I picked the attribute `index` to look up and check that I got the appropriate return value for each of the `AB`s contents. Would you like me to pick a different method? Add a doc string?